src/SSHDebugPS/ConnectionManager.cs

```diff
@@ -149,5 +174,140 @@ public static Connection GetSSHConnection(string name)
+            StringBuilder errorSB = new StringBuilder();
+            int exitCode = 0;
+
+            try
+            {
+                ManualResetEvent resetEvent = new ManualResetEvent(false);
+                commandRunner.ErrorOccured += ((sender, args) =>
+                {
+                    resetEvent.Set();
+                });
+
+                commandRunner.Closed += ((sender, args) =>
+                {
+                    exitCode = args;
+                    resetEvent.Set();
+                });
+
+                commandRunner.OutputReceived += ((sender, args) =>
+                {
+                    if (!string.IsNullOrWhiteSpace(args))
+                    {
+                        if (args.Trim()[0] != '{')
+                        {
+                            // output isn't json, command Error
+                            string errorMessage = string.Format(CultureInfo.CurrentCulture, UIResources.CommandExecutionErrorFormat, dockerCommand, args);
+                            throw new CommandFailedException(errorMessage);
+                        }
+
+                        var containerInstance = DockerContainerInstance.Create(args);
+                        if (containerInstance != null)
+                            containers.Add(containerInstance);
+                    }
+                });
+
+                commandRunner.Run();
+                resetEvent.WaitOne();
+
+                // might need to throw an exception here too??
+                if (exitCode != 0)
+                {
+                    Debug.Fail($"Exit Code: {exitCode}");
+                    return null;
+                }
+
+                return containers;
+            }
+            catch (Win32Exception ex)
+            {
+                // docker doesn't exist 
+                string errorMessage = string.Format(CultureInfo.CurrentCulture, UIResources.CommandExecutionErrorFormat, dockerCommand, ex.Message);
+                throw new CommandFailedException(errorMessage, ex);
+            }
+        }
+
+        public static IEnumerable<IContainerInstance> GetRemoteDockerContainers(IConnection connection)
+        {
+            SSHConnection sshConnection = connection as SSHConnection;
+            List<string> outputLines = new List<string>();
+            StringBuilder errorSB = new StringBuilder();
+            if (sshConnection == null)
+            {
+                return null;
+            }
+
+            List<DockerContainerInstance> containers = new List<DockerContainerInstance>();
+            RemoteCommandRunner commandRunner = new RemoteCommandRunner(dockerCommand, dockerPSArgs, sshConnection);
+
+            ManualResetEvent resetEvent = new ManualResetEvent(false);
+            int exitCode = 0;
+            commandRunner.ErrorOccured += ((sender, args) =>
+            {
+                resetEvent.Set();
```

Comment:
>resetEvent.Set(); [](start = 16, length = 17)

Should we be waiting to see if there is additional error output?

Reply:
we can do that.

---
In reply to: [287519036](https://github.com/microsoft/MIEngine/pull/868#discussion_r287519036) [](ancestors = 287519036)